### PR TITLE
Add positional docstring and pre/post conditions to defcached

### LINKED ISSUE #3: Make defcached more compatible with defn
- positional docstring
- pre/post conditions

### PR DESCRIPTION
**Pull request by fmnoise**
Closes https://github.com/fmnoise/coldbrew/issues/3

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -75,7 +75,19 @@ All meta passed to function name is preserved, so we can have private cached fun
       result)))
 ```
 
-*NB: Docstring is only supported as meta*
+Positional docstring as well as pre/post conditions map are also supported:
+```clojure
+(defcached fetch-customer-impl
+  "Fetches customer data by given id
```
Comment (Aeonax) [inline README.md]: Missed `"`

File: src/fmnoise/coldbrew.clj [hunk 2]
```diff
@@ -52,14 +52,29 @@
    [db date id]
    (query-customer-ltv db date id))
   "
-  [name args cache-key & fn-body]
-  (let [fname (gensym (str name "-cached-"))]
+  [name & fdecl]
```
Comment (Aeonax) [inline src/fmnoise/coldbrew.clj:55]: `fdecl`... What is it? 🤔😅  Maybe `body` will be enough? 😅

File: src/fmnoise/coldbrew.clj [hunk 3]
```diff
@@ -52,14 +52,29 @@
...
+                  (and docstring (map? ?prepost)) ?prepost
+                  (and (nil? docstring) (map? ?args)) ?args)
+        cache-key (cond
+                    (and docstring prepost) ?cache-key
+                    (and (nil? docstring) (nil? prepost)) ?args
+                    :else ?prepost)
+        fn-body (cond
+                  (and docstring prepost) (rest (rest (rest (rest fdecl))))
+                  (and (nil? docstring) (nil? prepost)) (rest (rest fdecl))
+                  :else (rest (rest (rest fdecl))))
```
Comment (Aeonax) [inline src/fmnoise/coldbrew.clj]: `(subvec fdecl 1-3)`?

**Review by Aeonax — APPROVED** [review]

File: src/fmnoise/coldbrew.clj (hunk 2 above)
Comment (fmnoise) [reply src/fmnoise/coldbrew.clj:55]: fdecl = function declaration, inspired by `defn` code from clojure.core
image

File: src/fmnoise/coldbrew.clj (hunk 3 above)
Comment (fmnoise) [reply src/fmnoise/coldbrew.clj]: image

File: README.md (hunk 1 above)
Comment (fmnoise) [reply README.md]: Good spot 👍🏽

File: src/fmnoise/coldbrew.clj (hunk 3 above)
Comment (fmnoise) [reply src/fmnoise/coldbrew.clj]: reworked with https://clojuredocs.org/clojure.core/nthrest

**Pull request merged**